internal/backends/winit/muda.rs

```diff
@@ -0,0 +1,98 @@
+use winit::window::Window;
+
+pub struct MudaAdapter {
+    entires: Vec<i_slint_core::items::MenuEntry>,
+    menubar: vtable::VBox<MenuVTable>,
+    // We need to keep menu alive, otherwise muda segfaults
+    _menu: muda::Menu,
+}
+
+impl MudaAdapter {
+    pub fn setup(
+        menubar: vtable::VBox<MenuVTable>,
+        winit_window: &Window,
+        proxy: EventLoopProxy<SlintUserEvent>,
+    ) -> Self {
+        type MenuRef<'a> = vtable::VRef<'a, MenuVTable>;
+        type EntryMap = Vec<i_slint_core::items::MenuEntry>;
+        fn generate_menu_entry(
+            menu: MenuRef,
+            entry: &i_slint_core::items::MenuEntry,
+            depth: usize,
+            map: &mut EntryMap,
+            window_id: &str,
+        ) -> Box<dyn muda::IsMenuItem> {
+            let id = muda::MenuId(format!("{window_id}|{}", map.len()));
+            map.push(entry.clone());
+            // the top level always has a sub menu regardless of entry.has_sub_menu
+            if !entry.has_sub_menu && depth != 0 {
+                Box::new(muda::MenuItem::with_id(
+                    id.clone(),
+                    &entry.title,
+                    true, /*entry.enabled*/
+                    None,
+                ))
+            } else {
+                let sub_menu =
+                    muda::Submenu::with_id(id.clone(), &entry.title, true /*entry.enabled*/);
+                if depth < 15 {
+                    let sub_entries = menu.sub_menu(Some(entry));
+                    for e in sub_entries {
+                        sub_menu
+                            .append(&*generate_menu_entry(menu, &e, depth + 1, map, window_id))
+                            .unwrap();
+                    }
+                } else {
+                    // infinite menu depth is possible, but we want to limit the amount of item passed to muda
+                    sub_menu
+                        .append(&muda::MenuItem::new(
+                            "<Error: Menu Depth limit reached>",
+                            false,
+                            None,
+                        ))
+                        .unwrap();
+                }
+                Box::new(sub_menu)
+            }
+        }
+
+        let menu = muda::Menu::new();
+        let mut map = EntryMap::new();
+        let menu_entries = menubar.sub_menu(None);
+        let window_id = u64::from(winit_window.id()).to_string();
+        for e in menu_entries {
+            menu.append(&*generate_menu_entry(menubar.borrow(), &e, 0, &mut map, &window_id))
+                .unwrap();
+        }
+
+        muda::MenuEvent::set_event_handler(Some(move |e| {
+            let _ = proxy.send_event(SlintUserEvent(crate::event_loop::CustomEvent::Muda(e)));
+        }));
+
+        #[cfg(target_os = "windows")]
+        {
+            use winit::raw_window_handle::*;
+            if let RawWindowHandle::Win32(handle) = winit_window.window_handle().unwrap().as_raw()
```

Comment:
This will be fun with the next winit version. We need to make sure that this function is only called *after* the window has been created, or else this `unwrap()` will panic. Would it be possible to ensure that this is called from within `ensure_window`?

Differently put: This looks like a "push" model, where the slint generated code invokes this, and I think in the future this will have to be a "pull" model, where the backend will fetch the information from the generated code when it's ready. I think it would be good to design it like that already now, since we know that we're going to have to do it that way in the future anyway.

(Not to stand in the way of this PR, but for your consideration)

Reply:
Yes, we can so that from show if the window adapter has a MudaAdapter or something like that.

Reply:
Cool, that also works (fingers crossed). Good idea.